### PR TITLE
feat: add Stooge sort algorithm

### PR DESCRIPTION
**Pull request by mindaugl**
#### Checklist

- [ ] Added description of change
- [ ] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [ ] Added tests and example, test must pass
- [ ] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [ ] Relevant documentation/comments is changed or added
- [ ] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [ ] Search previous suggestions before making a new one, as yours may be a duplicate.
- [ ] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

File: sorting/stooge_sort.cpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ *	- sort first two thirds of the array
+ *	- sort last two thirds of the array
+ *  - sort first two thirds of the array
+ * It's time complexity is O(n^(log3/log1.5)), which is about O(n^2.7), 
+ * which makes it to be not the most efficient sorting algorithm 
+ * on the street on average. Space complexity is O(1).
+ */
+
+#include <vector>
+#include <iostream>
```
Comment (Panquesito7) [inline sorting/stooge_sort.cpp:17]: Please add a one-line description of what the library/header is for (**see the example below**).

```c++
#include <iostream>    /// for IO operations
#include <cassert>    /// for assert
```

File: sorting/stooge_sort.cpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+		if (j - i > 1) {
+			int third = (j - i + 1) / 3;
+			stoogeSort(L, i, j - third);
+			stoogeSort(L, i + third, j);
+			stoogeSort(L, i, j - third);
+		}
+	}
+	return;
+}
+
```
Comment (Panquesito7) [inline sorting/stooge_sort.cpp:33]: Please add self-test cases so we can verify your program works properly (**see the [typical structure of a program](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) as a reference**). Let us know if you need any help here or on our [**Discord server**](https://the-algorithms.com/discord). 😃

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please fix the [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/actions/runs/4393387650/jobs/7702649789) warnings.
Let us know if you need any help with that. 🙂

File: sorting/stooge_sort.cpp (hunk 2 above)
Comment (mindaugl) [reply sorting/stooge_sort.cpp:41]: Just added documentation and 3 test cases, thanks!

File: sorting/stooge_sort.cpp (hunk 1 above)
Comment (mindaugl) [reply sorting/stooge_sort.cpp]: Just added comments for includes

File: sorting/stooge_sort.cpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+/**
+ * @brief Function to test sorting algorithm
```
Comment (Panquesito7) [inline sorting/stooge_sort.cpp:43]:
```suggestion
 * @brief Function to test sorting algorithm
 * @returns void
```

File: sorting/stooge_sort.cpp [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+ */
+void test1() {
+	std::vector<int> L = { 8, 9, 10, 4, 3, 5, 1 };
+	stoogeSort(&L, 0, L.size() - 1);
+	assert(std::is_sorted(std::begin(L), std::end(L)));
+}
+
+/**
+ * @brief Function to test sorting algorithm, one element
```
Comment (Panquesito7) [inline sorting/stooge_sort.cpp:52]:
```suggestion
 * @brief Function to test sorting algorithm, one element
 * @returns void
```

File: sorting/stooge_sort.cpp [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+ */
+void test2() {
+	std::vector<int> L = { -1};
+	stoogeSort(&L, 0, L.size() - 1);
+	assert(std::is_sorted(std::begin(L), std::end(L)));
+}
+
+/**
+ * @brief Function to test sorting algorithm, repeating elements
```
Comment (Panquesito7) [inline sorting/stooge_sort.cpp:61]:
```suggestion
 * @brief Function to test sorting algorithm, repeating elements
 * @returns void
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
These are needed so that Doxygen detects the functions properly.

File: sorting/stooge_sort.cpp (hunk 3 above)
Comment (mindaugl) [reply sorting/stooge_sort.cpp:43]: Updated

File: sorting/stooge_sort.cpp (hunk 5 above)
Comment (mindaugl) [reply sorting/stooge_sort.cpp:63]: Updated

File: sorting/stooge_sort.cpp (hunk 4 above)
Comment (mindaugl) [reply sorting/stooge_sort.cpp:53]: Updated

File: sorting/stooge_sort.cpp [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+ *	- sort last two thirds of the array
+ *  - sort first two thirds of the array
+ * It's time complexity is O(n^(log3/log1.5)), which is about O(n^2.7),
+ * which makes it to be not the most efficient sorting algorithm
+ * on the street on average. Space complexity is O(1).
+ */
+
+#include <vector> /// for vector
+#include <cassert> /// for assert
+#include <algorithm>  /// for std::is_sorted
```
Comment (Panquesito7) [inline sorting/stooge_sort.cpp:18]:
```suggestion
#include <vector>     /// for vector
#include <cassert>    /// for assert
#include <algorithm>  /// for std::is_sorted
#include <iostream>   /// for IO operations
```

File: sorting/stooge_sort.cpp [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+/**
+ * @brief Main function
+ * @param argc commandline argument count (ignored)
+ * @param argv commandline array of arguments (ignored)
+ * @returns 0 on exit
+ */
+int main() {
+	test1();
+	test2();
+	test3();
```
Comment (Panquesito7) [inline sorting/stooge_sort.cpp:81]: With no printing message, the user might be confused on what the program's doing or did.
```suggestion
	test3();

	std::cout << "All tests have successfully passed!\n";
```

Comment (mindaugl) [reply sorting/stooge_sort.cpp:82]: Done

**Review by Panquesito7 — APPROVED** [review]
Looks great. Thanks! 🚀

File: sorting/stooge_sort.cpp [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+ /**
+  * @brief The stoogeSort() function is used for sorting the array.
+  * @param L - vector of values (int) to be sorted in in place (ascending order)
+  * @param i - the first index of the array (0)
+  * @param j - the last index of the array (L.size() - 1)
+  * @returns void
+  */
+void stoogeSort(std::vector<int>* L, size_t i, size_t j) {
+	if (i < j) {
```
Comment (alexpantyukhin) [inline sorting/stooge_sort.cpp:29]: What if to reverse:
```cpp
if (i >= j){
    return;
}
```

File: sorting/stooge_sort.cpp [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+			std::swap((*L)[i], (*L)[j]);
+		}
+		if (j - i > 1) {
+			size_t third = (j - i + 1) / 3;
+			stoogeSort(L, i, j - third);
+			stoogeSort(L, i + third, j);
+			stoogeSort(L, i, j - third);
+		}
+	}
+	return;
```
Comment (tjgurwara99) [inline sorting/stooge_sort.cpp:40]: This is logically correct but could you think of a different way to write this? `@alexpantyukhin` gave the correct suggestion but you've marked it as resolved without considering its impact.

```cpp
void stoogeSort(std::vector<int>* L, size_t i, size_t j) {
 	if (i >= j) {
		return;
	}
	if ((*L)[i] > (*L)[j]) {
 		std::swap((*L)[i], (*L)[j]);
 	}
 	if (j - i > 1) {
 		size_t third = (j - i + 1) / 3;
 		stoogeSort(L, i, j - third);
 		stoogeSort(L, i + third, j);
 		stoogeSort(L, i, j - third);
 	}
}
```

Just adding the guard clause

```cpp
 	if (i >= j) {
		return;
	}
```

makes so much difference in nesting. The use of a guard clause makes reading your code easier for people reviewing your code, and remember that code is read more often than written therefore you should write it in a way to make this easier for readers.

**Review by tjgurwara99 — CHANGES_REQUESTED** [review]

**Review by Panquesito7 — APPROVED** [review]

**Review by tjgurwara99 — APPROVED** [review]

**Pull request merged**